File: submit_20176342_hw5/P2POmokServer.java
package submit_20176342_hw5;
/**
 * 20176342 Song Min Joon
 * P2POmokServer.java
 **/

import java.util.*;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.UnknownHostException;
import java.sql.Time;

public class P2POmokServer {
	class client {
		String nickname;
		int uniqueID;
		Socket conn;
		String ip;
		String port;
		String remote;
	}

    final public static int serverPort = 26342;// for server port
    public static int uniqueID = 1;// for server port
	public static HashMap<Integer ,client> clientMap = new HashMap<Integer, client>();

	public static void main(String args[]){
        ServerSocket listener = null;
        Socket conn = null;
        try {
            listener = new ServerSocket(serverPort); // create listener for tcp socket

        } catch (IOException e) {

        }
        System.out.printf("Server is ready to receive on port %s\n", serverPort);
        Runtime.getRuntime().addShutdownHook(new ByeByeThread(listener)); // shutdown hook for graceful exit

        try {
            while (true) {
                // listener is waiting for tcp connection of clients.
                conn = listener.accept();
                System.out.printf("Connection request from %s\n", conn.getInetAddress() + ":" + conn.getPort());

                // when client is connect to server, make individual sub-thread to communicate
                // each client.
                ServerReceiver th = new ServerReceiver(conn);
                th.start();
            }

        } catch (IOException e) {

        }
	}
	public static void byebye() {
        // print bye bye~
        System.out.println("Bye bye~");
    }

    public static int registerClient(client client, int uniqueID){
        clientMap.put(uniqueID, client);
        return uniqueID;
    }
    public static void unregisterClient(int uniqueID){
        clientMap.remove(uniqueID);
    }
	public static class ByeByeThread extends Thread {
        // ByeBye Thread for graceful exit program.
        ServerSocket listener;

        ByeByeThread(ServerSocket listener) {
            this.listener = listener;
        }

        public void run() {
            try {
                Thread.sleep(200);
                byebye();//print bye bye~

                try {
                    this.listener.close(); // close listener
                } catch (IOException e) {

                }

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                e.printStackTrace();
            }
        }
    }

	static class ServerReceiver extends Thread {
        // ServerReceiver Thread for each client
        Socket conn;
        DataInputStream is;
        DataOutputStream os;

        ServerReceiver(Socket conn) {
            this.conn = conn;
            try {
                is = new DataInputStream(conn.getInputStream());
                os = new DataOutputStream(conn.getOutputStream());
            } catch (IOException e) {

            }
        }

        public void run() {
            while (true) {
                try {
                    int bufferSize = 1024;
                    byte[] buffer = new byte[bufferSize];

                    int count = is.read(buffer);

                    if (count == -1) {
                        System.out.println("Client disconnected");
                        break;
                    }

                    // read packet from server and handle message
                    String clientMsg = new String(buffer, 0, count);

                    handleMsg(clientMsg);

                } catch (IOException e) {
                    System.out.println("client is disconnected");
                    break;
                }
            }

        }

        public void handleMsg(String msg) {

            /*
             * client packet form
             * 
             * (Option Number|Message)
             * 
             * 1|blah blah blah...
             * 1|hello world!
             * 
             * 2| => message is not required
             * 3| => message is not required
             * 4| => message is not required
             * 
             * 5| => maybe not arrived??
             * 
             */

            String[] req = msg.split("\\|"); // split client packet by '|' and takes option and convert to Integer.
            int requestOption = Integer.parseInt(req[0]);

            System.out.printf("Command %d\n\n\n", requestOption); // print Command #

            String requestData;
            try {
                requestData = req[1]; // get message parameter from packet.

            } catch (ArrayIndexOutOfBoundsException e) {
                requestData = "";
            }

            switch (requestOption) {
                case 1:
                    // Option 1
                    sendPacket(requestData.toUpperCase());
                    break;
                case 2:
                    // Option 2
                    String ip = (this.conn.getInetAddress() + ":" + this.conn.getPort());
                    sendPacket(ip);
                    break;
                case 3:
                    // Option 3
                    // sendPacket(totalRequest + "");
                    break;
                case 4:
                    // Option 4
                    // sendPacket(milliToTimeFormat(elapsed) + "");
                    break;
                case 5:
                    // Option 5
                    try {
                        this.conn.close();
                        this.stop();
                    } catch (IOException e) {

                    }
                    break;
                default:
                    // Option default
                    try {
                        this.conn.close();
                        this.stop();
                    } catch (IOException e) {

                    }
                    break;
            }
        }

        void sendPacket(String packet) {
            // send packet to server
            try {
                os.write(packet.getBytes());
                os.flush();
            } catch (IOException e) {

            }

        }

    }
}

File: submit_20176342_hw5/P2POmokServer.go
/**
 * 20176342 Song Min Joon
 * P2POmokServer.go
 **/
package main

import (
	"fmt"
	"net"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"
)

var serverPort string = "56342" // for server port
var uniqueID int = 1
var clientMap map[int]client

type client struct {
	nickname string
	uniqueID int
	conn     net.Conn
	ip       string
	port     string
	remote   string
}

func main() {
	c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM) // for exit the program gracefully
	go func() {
		for sig := range c {
			// sig is a ^C, handle it
			_ = sig
			byebye() // if this program is interrupt by Ctrl-c, print Bye bye and exits gracefully
		}
	}()

	listener, _ := net.Listen("tcp", ":"+serverPort)
	fmt.Printf("Server is ready to receive on port %s\n", serverPort)

	clientMap = make(map[int]client) //make client map for record clients

	for {
		//listener is waiting for tcp connection of clients.
		conn, err := listener.Accept()
		if err != nil {
			handleError(conn, err, "server accept error..")
		}


		nickBuffer := make([]byte, 1024)

		count, err := conn.Read(nickBuffer)

		if err != nil || count == 0 {
			continue
			return
		}

		targetNick := string(nickBuffer[:count])



		remoteAddr := conn.RemoteAddr().String()
		lastIdx := strings.LastIndex(remoteAddr, ":")

		newIP := remoteAddr[0:lastIdx]
		newPort := remoteAddr[lastIdx+1:]

		fmt.Printf("%s joined from from %s. UDP port %s\n", targetNick,conn.RemoteAddr().String(),newPort)

		newClient := client{targetNick, uniqueID, conn, newIP, newPort, remoteAddr}

		registerClient(newClient, uniqueID)

		go handleMsg(newClient, uniqueID) // when client is connect to server, make go-routine to communicate with client.
		uniqueID++

		if len(clientMap) == 1 {
			sendPacket(newClient, "waiting||")
			fmt.Printf("1 user connected, waiting for another\n")
		} else if len(clientMap) == 2 {
			var otherRemote client
			for _, v := range clientMap {
				if v.uniqueID != newClient.uniqueID {
					otherRemote = v
					sendPacket(v, "matched|"+newClient.nickname+"|"+newClient.remote)
				}
			}
			sendPacket(newClient, "success|"+otherRemote.nickname+"|"+otherRemote.remote)

			fmt.Printf("2 users connected, notifying %s and %s\n",otherRemote.nickname,newClient.nickname )


			for _, v := range clientMap {
				unregisterClient(v.uniqueID)
			}
			fmt.Printf("%s and %s disconnected.\n",otherRemote.nickname,newClient.nickname )

		}

	}

	defer byebye() // although when client gets panic, defer should disconnect socket gracefully

}

func handleMsg(client client, cid int) {
	for {
		buffer := make([]byte, 1024)

		count, err := client.conn.Read(buffer)

		//when client sends packet

		if err != nil {
			unregisterClient(cid)
			// broadCastToAll(6, fmt.Sprintf("%s is disconnected. There are %d users in the chat room", client.nickname, len(clientMap)))
			if len(clientMap) == 1{
				fmt.Printf("%s disconnected. 1 User left in server.\n",client.nickname)
			} else if len(clientMap) == 0{
				fmt.Printf("%s disconnected. No User left in server.\n",client.nickname)
			}
			handleError(client.conn, err, "%s disconnected!")
			return
		}

		if count == 0 {
			// fmt.Printf("return ! \n")
			return
		}
	}
}
func registerClient(client client, uniqueID int) int {
	clientMap[uniqueID] = client
	return uniqueID
}

func unregisterClient(uniqueID int) {
	if _, ok := clientMap[uniqueID]; ok {
		delete(clientMap, uniqueID)
	}
}

func byebye() {
	fmt.Printf("Bye bye~\n")
	os.Exit(0)
}

func sendPacket(client client, serverMsg string) {
	//send packet to client
	time.Sleep(time.Millisecond * 1) // minimum delay to deliver packet to client.
	client.conn.Write([]byte(serverMsg))
	fmt.Printf("send packet %s => %s \n", client.nickname, serverMsg)
}

func handleError(conn net.Conn, err error, errmsg string) {
	//handle error and print
	if conn != nil {
		conn.Close()
	}
}
